Bluetooth: AVRCP: implementation of Get Capabilities command.

gzh-terry opened the pull request:
This patch allows the upper layer to acquire the remote AVRCP capabilities. These capabilities can be further used in register notifications.

gzh-terry commented:
Hi @makeshi ,

Would you like to take a look and tell if the changes are acceptable to you?

gzh-terry commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -132,6 +157,12 @@ struct bt_avrcp_passthrough_rsp {
 };
 
+struct bt_avrcp_get_capabilities_rsp {
+	uint8_t capability_id;     /**< bt_avrcp_cap_t */
```
> In the header file, already define bt_avrcp_cap_t, why not use bt_avrcp_cap_t?

To make the structure smaller. `bt_avrcp_cap_t` is a enum and usally takes more than 1 byte.

jhedberg replied:
You can use `__packed` with enums if you need to compress them to the smallest possible integer type.

gzh-terry replied:
Changed accordingly :smile:

gzh-terry commented on subsys/bluetooth/host/classic/shell/avrcp.c:
```diff
@@ -217,13 +238,40 @@ static int cmd_pause(const struct shell *sh, int32_t argc, char *argv[])
+	if (!strcmp(cap_id, "company")) {
+		bt_avrcp_get_capabilities(default_avrcp, BT_AVRCP_CAP_COMPANY_ID);
+	} else if (!strcmp(cap_id, "event")) {
+		bt_avrcp_get_capabilities(default_avrcp, BT_AVRCP_CAP_EVENTS_SUPPORTED);
```
Can change it later in another PR, since there are many similar cases in this file.

gzh-terry replied:
> It is better change return value of bt_avrcp_get_capabilities().

Can change it later in another PR, since there are many similar cases in this file.

gzh-terry commented on subsys/bluetooth/host/classic/shell/avrcp.c:
```diff
@@ -43,6 +43,26 @@ static void avrcp_disconnected(struct bt_avrcp *avrcp)
+		for (i = 0; i < rsp->capability_count; i++) {
+			bt_shell_print("Remote supported EventID = 0x%02x", rsp->capability[i]);
+		}
+		break;
```
> Need add default error handle here?

Shell is not really a complete application, but a tool to verify the functions of the Host stack. I think the existing messages are enough for this purpose.

BTW, invalid values are detected at `process_get_capabilities_rsp` in `subsys/bluetooth/host/classic/avrcp.c`

gzh-terry commented on subsys/bluetooth/host/classic/avrcp_internal.h:
```diff
@@ -37,6 +37,67 @@ typedef enum __packed {
+	BT_AVRCP_PDU_ID_GET_ELEMENT_ATTRIBUTES = 0x20,
+
+	/** Notifications */
+	BT_AVRCP_PDU_ID_GET_PLAY_STATUS = 0x30,
```
Hi @MarkWangChinese 

It cames here:
![image](https://github.com/user-attachments/assets/f3f6682e-aa46-498d-8238-7712fb71bbb7)

MarkWangChinese replied:
OK, Thanks @gzh-terry

gzh-terry commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -243,22 +243,101 @@ static void avrcp_disconnected(struct bt_avctp *session)
+		return;
+	}
+
+	pdu_id = buf->data[0];
```
> why not `net_buf_pull_u8`?

To keep the first byte (PDU ID) stay in `buf`
This leaves the entire payload to the next stage, i.e., `process_get_capabilities_rsp()`

The previous operations, i.e.,
```
	avrcp_hdr = net_buf_pull_mem(buf, sizeof(*avrcp_hdr));
	company_id = net_buf_pull_be24(buf);
```
however, is to remove the AV/C header from the response, which not belongs to (and not necessary at) the next stage.

MarkWangChinese replied:
From my understand, this `avrcp_vendor_dependent_handler` function is used to distribute different vendor dependent `pdu_id` to different dedicated handler functions. This `avrcp_vendor_dependent_handler` should consume all the common data for all the vendor dependent pdu, then the dedicated vendor dependent handler function only care its dedicated data in the buf. This is my understand.

gzh-terry replied:
> From my understand, this `avrcp_vendor_dependent_handler` function is used to distribute different vendor dependent `pdu_id` to different dedicated handler functions. This `avrcp_vendor_dependent_handler` should consume all the common data for all the vendor dependent pdu, then the dedicated vendor dependent handler function only care its dedicated data in the buf. This is my understand.

Good reason, I've changed it as you said.

gzh-terry commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -243,22 +243,101 @@ static void avrcp_disconnected(struct bt_avctp *session)
+	}
+
+	net_buf_pull_u8(buf); /**< PDU ID 0x10 - get capabilities */
+	net_buf_pull_u8(buf); /**< Packet Type 0x0 - single frame*/
```
> Is it OK to put them in `avrcp_vendor_dependent_handler` for common process.

The first octet "PDU ID" is move to the general process.
The second octet "Packet Type" remains here, otherwise we might need a process_xxx_segment in the future.
(Although it's not possble for this command)

lylezhu2012 replied:
If the structure `AVRCP specific AV/C PDU format` used here, I think we should not have such question.

gzh-terry replied:
Done.

gzh-terry committed:
Bluetooth: AVRCP: implementation of Get Capabilities command.

This patch allows the upper layer to acquire the remote AVRCP
capabilities. These capabilities can be further used in register
notifications.

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -243,22 +243,100 @@ static void avrcp_disconnected(struct bt_avctp *session)
+	uint32_t company_id;
+	uint8_t pdu_id;
+
+	if (buf->len < 7) {
```
Suggest change to 
`#define COMPANY_ID_SIZE  3`
`if (buf->len < (sizeof(*avrcp_hdr) + COMPANY_ID_SIZE + sizeof(pdu_id))) {`

gzh-terry replied:
Updated.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -589,6 +688,26 @@ static int avrcp_send(struct bt_avrcp *avrcp, struct net_buf *buf)
+int bt_avrcp_get_capabilities(struct bt_avrcp *avrcp, bt_avrcp_cap_t capability_id)
+{
+	struct net_buf *buf;
+	uint8_t param[5];
```
There is a definition of `AVRCP specific AV/C PDU format` defined by AVRCP specific. Suggest use a struct to pack the PDU. It is easy to read.

gzh-terry replied:
Done.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -589,6 +688,26 @@ static int avrcp_send(struct bt_avrcp *avrcp, struct net_buf *buf)
+	net_buf_add_be24(buf, BT_AVRCP_COMPANY_ID_BLUETOOTH_SIG);
+	param[0] = BT_AVRCP_PDU_ID_GET_CAPABILITIES;
+	param[1] = 0x00;
+	param[2] = 0x00;
+	param[3] = 0x01;
```
Using structures helps enhance readability.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -309,12 +388,12 @@ static void avrcp_pass_through_handler(struct bt_avrcp *avrcp, struct net_buf *b
 	struct bt_avrcp_passthrough_rsp rsp;
 	uint8_t tmp;
 
+	avrcp_hdr = net_buf_pull_mem(buf, sizeof(*avrcp_hdr));
+
```
It seams the change is not related to this PR. Right?

gzh-terry replied:
Hi @lylezhu2012 

This is to make `avrcp_hdr` accessable for both commands and responses.

I've noticed that @makeshi has work on it.

The other changes are to keep the style consistent.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -351,8 +430,7 @@ static int avrcp_recv(struct bt_avctp *session, struct net_buf *buf)
 
-	avctp_hdr = (void *)buf->data;
-	net_buf_pull(buf, sizeof(*avctp_hdr));
+	avctp_hdr = net_buf_pull_mem(buf, sizeof(*avctp_hdr));
```
same here.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -278,11 +356,12 @@ static void avrcp_subunit_info_handler(struct bt_avrcp *avrcp, struct net_buf *b
 	struct bt_avrcp_subunit_info_rsp rsp;
 	uint8_t tmp;
 
+	avrcp_hdr = net_buf_pull_mem(buf, sizeof(*avrcp_hdr));
```
same here.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -243,22 +243,100 @@ static void avrcp_disconnected(struct bt_avctp *session)
 	struct bt_avrcp_header *avrcp_hdr;
 	struct bt_avrcp_unit_info_rsp rsp;
 
+	avrcp_hdr = net_buf_pull_mem(buf, sizeof(*avrcp_hdr));
```
same here.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -243,22 +243,100 @@ static void avrcp_disconnected(struct bt_avctp *session)
+	uint16_t length;
+	uint16_t expected_length;
+
+	if (buf->len < 3) {
```
Suggest use a struct to handle the response instead of using a hard code `3` here.

gzh-terry replied:
Done.

lylezhu2012 commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -125,13 +150,19 @@ struct bt_avrcp_subunit_info_rsp {
 	const uint8_t *payload;
 };
 
+struct bt_avrcp_get_capabilities_rsp {
+	bt_avrcp_cap_t capability_id;
+	uint8_t capability_count;  /**< number of items contained in *capability */
+	const uint8_t *capability; /**< 1 or 3 octets each depends on capability_id */
+};
```
Suggest change to 
```
struct bt_avrcp_get_capabilities_rsp {
	uint8_t capability_id; /**< see @ref bt_avrcp_cap_t */
	uint8_t capability_count;  /**< number of items contained in *capability */
	const uint8_t capability[]; /**< 1 or 3 octets each depends on capability_id */
} __packed;
```

jhedberg replied:
Are these structs supposed to represent local state or map to the raw protocol data? If the latter, then fixed size integer types (e.g. `uint8_t`) and `__packed` are indeed correct. However, looks like there's inconsistency here e.g. with `bt_avrcp_passthrough_rsp` right above, since it has a `const uint8_t *payload` pointer.

lylezhu2012 replied:
The structure format I suggested is defined by specific. It maps to the raw protocol data.

gzh-terry replied:
> The structure format I suggested is defined by specific. It maps to the raw protocol data.

The original design of these structures is NOT a map of the raw data (See `bt_avrcp_passthrough_rsp`)
I think it would be a burden for applications to decode the raw AVRC frame.

Alternatively, the structure would be:
```
struct bt_avrcp_passthrough_rsp {
	uint8_t reserved: 4;
	uint8_t response: 4;
	uint8_t subunit_type: 5;
	uint8_t subunit_id: 3;
	uint8_t opcode;
	uint8_t state_flag: 1;
	uint8_t operation_id: 7;
	uint8_t len;
	uint8_t data[0];
} __packed;
```

jhedberg replied:
Be careful with bitfields in packed structs. The in-memory layout will be different depending on the endianness of the architecture, and so you have to define two versions of the structure depending on this. What I've preferred to do in cases like this is to just use complete `uint8_t` fields and instead have helper macros to access specific subsets of bits.

gzh-terry replied:
> Be careful with bitfields in packed structs. The in-memory layout will be different depending on the endianness of the architecture, and so you have to define two versions of the structure depending on this. What I've preferred to do in cases like this is to just use complete `uint8_t` fields and instead have helper macros to access specific subsets of bits.

As a trade-off, I think the current version would be better:
```
struct bt_avrcp_passthrough_rsp {
	bt_avrcp_rsp_t response;
	bt_avrcp_opid_t operation_id;
	bt_avrcp_button_state_t state;
	uint8_t len;
	const uint8_t *payload;
};

struct bt_avrcp_get_capabilities_rsp {
	bt_avrcp_cap_t capability_id;
	uint8_t capability_count;  /**< number of items contained in *capability */
	const uint8_t *capability; /**< 1 or 3 octets each depends on capability_id */
};
```

What do you think? @lylezhu2012

jhedberg replied:
As long as you're natively constructing those structs, that should be ok. However if you're applying a structure pointer on a protocol buffer (e.g. net_buf payload) you should use fixed-size integer types and __packed.

jhedberg replied:
To be clear: I'm fine either way. What I don't want is some kind of hybrid option, since I think it's important to use the same conventions and patters throughout the Bluetooth host stack codebase. The existing convention for raw protocol data structs is to use the fixed-size integer types (`uint8_t`, `uint16_t`, etc) together with `__packed`, and you generally wouldn't find any custom typedefs in such structs.

gzh-terry replied:
Updated
```C
struct bt_avrcp_passthrough_rsp {
	uint8_t byte0; /**< [7]: state_flag, [6:0]: opid */
	uint8_t data_len;
	uint8_t data[];
};

struct bt_avrcp_get_cap_rsp {
	uint8_t cap_id;  /**< bt_avrcp_cap_t */
	uint8_t cap_cnt; /**< number of items contained in *cap */
	uint8_t cap[];   /**< 1 or 3 octets each depends on cap_id */
};
```

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -243,22 +243,100 @@ static void avrcp_disconnected(struct bt_avctp *session)
+	net_buf_pull_u8(buf); /**< Packet Type 0x0 - single frame*/
+	length = net_buf_pull_be16(buf);
+
+	if (length < 2 || length < buf->len) {
```
Suggest change to `if (length < sizeof(*rsp) || length != buf->len) {`

gzh-terry replied:
Done.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -243,22 +243,100 @@ static void avrcp_disconnected(struct bt_avctp *session)
 
+static void process_get_capabilities_rsp(struct bt_avrcp *avrcp, struct net_buf *buf)
+{
+	struct bt_avrcp_get_capabilities_rsp rsp;
```
Suggest change to `struct bt_avrcp_get_capabilities_rsp *rsp;`

gzh-terry replied:
Done.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -243,22 +243,100 @@ static void avrcp_disconnected(struct bt_avctp *session)
+		return;
+	}
+
+	rsp.capability_id = net_buf_pull_u8(buf);
+	rsp.capability_count = net_buf_pull_u8(buf);
+	rsp.capability = buf->data;
```
rsp = net_buf_pull_mem(buf, sizeof(*rsp));

gzh-terry replied:
Done.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -243,22 +243,100 @@ static void avrcp_disconnected(struct bt_avctp *session)
+	rsp.capability_count = net_buf_pull_u8(buf);
+	rsp.capability = buf->data;
+
+	switch (rsp.capability_id) {
+	case BT_AVRCP_CAP_COMPANY_ID:
+		expected_length = rsp.capability_count * 3;
+		break;
+	case BT_AVRCP_CAP_EVENTS_SUPPORTED:
+		expected_length = rsp.capability_count;
+		break;
+	default:
+		LOG_ERR("Unrecognized capability = 0x%x", rsp.capability_id);
+		return;
+	}
+
+	if (buf->len < expected_length) {
+		LOG_ERR("Invalid capability payload length: %d, expected: %d", buf->len,
+			expected_length);
+		return;
+	}
+
+	avrcp_cb->get_capabilities_rsp(avrcp, &rsp);
```
I think the capability id of response should be checked to make sure it is aligned with the get capability command.
Otherwise, the GetCapabilities response for CompanyID is required by application, but the GetCapabilities response for EVENTS_SUPPORTED is reported. It is not seasonable.

gzh-terry replied:
> I think the capability id of response should be checked to make sure it is aligned with the get capability command. Otherwise, the GetCapabilities response for CompanyID is required by application, but the GetCapabilities response for EVENTS_SUPPORTED is reported. It is not seasonable.

I think is fine at the app side since there is a `capability_id` in `struct bt_avrcp_get_capabilities_rsp`

Not really necessary to add a check when it is not an error.

Likewise, we would have to cache all the commands that have been sent to make sure the responses are correctly corresponding to a command. These commands, however, can take memories for a long lifetime (take `RegisterNotifications` for example).

Some other error detections, like the check in data length, are necessary because it would lead to severe issue (like memory overflow).

gzh-terry replied:
> it should be only one command is ongoing

We can not have this assumption.

And the buffered command is only used for timeout check. We can remove it if such a check is not necessary at the Host stack.

Verifing each response requires the entire list of the pending command, which is not ideal in our experience (There will be many RegistereNotificationCommands on pending).

And an unsolicited response, however, does not really cause a issue that can not handled by the upper layer.

lylezhu2012 replied:
> > it should be only one command is ongoing
> 
> We can not have this assumption.

How it confirm that the buffer will not be overwritten by the following command. 
Then there are two problems.
Problem 1: The overwritten command cannot be managed by the timeout mechanism. If the peer does not reply, the command will never end.
Problem 2: If the peer replies, the received response will be discarded as invalid information because the cached information is overwritten.

gzh-terry replied:
Hi @lylezhu2012 

In the current implementation, the buffer **will be overwritten**.

As for issue 1: It is not really necessary to monitor the timeout for all commands. Indeed the timeout only lead to a print. If any actions are required, the application typlically has its own timeout (and the time duration can be tailored according to the product definition).

As for issue 2: The response will not be discarded. Nothing happens with the cached information except a log printed.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -243,22 +243,100 @@ static void avrcp_disconnected(struct bt_avctp *session)
+
+	switch (rsp.capability_id) {
+	case BT_AVRCP_CAP_COMPANY_ID:
+		expected_length = rsp.capability_count * 3;
```
Suggest change to `expected_length = rsp.capability_count * COMPANY_ID_SIZE;`

gzh-terry replied:
Updated.

gzh-terry committed:
Bluetooth: AVRCP: use spec defined structures on APIs.

This patch replaced the structures defined by zephyr APIs
by those defined by AVRCP Spec.

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

jhedberg commented on subsys/bluetooth/host/classic/avrcp_internal.h:
```diff
@@ -49,8 +118,16 @@ struct bt_avrcp_header {
+struct bt_avrcp_avc_pdu {
+	uint8_t pdu_id;
+	uint8_t byte1; /** [7:2]: Reserved, [1:0]: Packet Type */
+	uint8_t parameter_length_high;
+	uint8_t parameter_length_low;
```
Shouldn't these be just one `uint16_t` value?

gzh-terry replied:
> Shouldn't these be just one `uint16_t` value?

To avoid big endian issue.

`uint16_t` typically have the first byte to be LSB.

gzh-terry replied:
@jhedberg I used to have a `uint16_t` in my first try :smile:

jhedberg replied:
Why did you change it?

jhedberg replied:
> `uint16_t` typically have the first byte to be LSB.

When you have `__packed` structs, the multi-byte integer fields in it are (by convention) always assumed to use protocol endianness. To access them, you always need to use functions like `sys_le16_to_cpu()`.

Unrelated to this PR: It'd be nice if Zephyr had similar byte-order specific integer types that Linux has, which use the gcc `__bitwise` feature. Then we could have types like `le16_t length;` in these structs, and the conversion functions would look like:
```c
uint16_t sys_le16_to_cpu(le16_t val);
```

Anyway, since we don't have such types (yet), `__packed` structs, at least in a Bluetooth context, are always to contain protocol-endian multi-byte fields.

jhedberg replied:
> I was just a little bit worried that someone may forget to use `sys_le16_to_cpu()` or `sys_be16_to_cpu()` in the future.

That's indeed a fair worry, and something that would be solved by adopting `__bitwise` types. That said, this has always been the convention in Zephyr's Bluetooth code, so it's nothing unique for e.g. AVRCP.

jhedberg commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -589,6 +689,26 @@ static int avrcp_send(struct bt_avrcp *avrcp, struct net_buf *buf)
+	net_buf_add_be24(buf, BT_AVRCP_COMPANY_ID_BLUETOOTH_SIG);
+	pdu.pdu_id = BT_AVRCP_PDU_ID_GET_CAPABILITIES;
+	BT_AVRCP_AVC_PDU_SET_PACKET_TYPE(&pdu, BT_AVRVP_PKT_TYPE_SINGLE);
+	pdu.parameter_length = sys_cpu_to_be16(0x0001);
```
`pdu.param_len = sys_cpu_to_be16(sizeof(cap_id))`

gzh-terry replied:
> `pdu.param_len = sys_cpu_to_be16(sizeof(cap_id))`

`bt_avrcp_cap_t` is not explicitly one byte (although it is now), can we use `sizeof(uiint8_t)` instead?

jhedberg replied:
Because of this it's a bit questionable to use enums with protocol structs that require fixed sized fields. I'd either propose to use defines, like we do in most other places of the Zephyr Bluetooth stack, or then add a `BUILD_ASSERT(sizeof(bt_avrcp_cap_t) == 1)`

jhedberg replied:
You could also just use `uint8_t cap_id` for the variable definition.

gzh-terry replied:
> You could also just use `uint8_t cap_id` for the variable definition.

This sounds reasonable, already updated.

jhedberg commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -124,12 +149,21 @@ struct bt_avrcp_subunit_info_rsp {
+};
+
+struct bt_avrcp_get_capabilities_rsp {
+	uint8_t capability_id;    /**< bt_avrcp_cap_t */
+	uint8_t capability_count; /**< number of items contained in *capability */
+	uint8_t capability[0];    /**< 1 or 3 octets each depends on capability_id */
```
Could you just use more concise names for the structure members. I think it's going to be equally clear if you call these `cap_id`, `cap_count` and `cap`. Same for the structure name actually, and all other places you're spelling out the full "capability". Using shorter names makes the lines shorter, causing you to need less line splits, and thereby makes the code more readable.

gzh-terry replied:
Sure, will update.

gzh-terry replied:
Done.

jhedberg commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -124,12 +149,21 @@ struct bt_avrcp_subunit_info_rsp {
-	uint8_t len;
-	const uint8_t *payload;
+	uint8_t byte0; /**< [7]: state_flag, [6:0]: operation_id */
+	uint8_t operation_data_field_length;
+	uint8_t operation_data[0];
```
You really seem to enjoy long names :)

Could you just follow the established naming convention in Zephyr's Bluetooth stack to use "len" for length, "op" for "operation", and "param" for "parameter". This could e.g. become:

```c
	uint8_t data_len;
	uint8_t data[0];
```

Oh, talking about `op[0]`, you should be aware of #81895.

gzh-terry replied:
I was trying to use the original name from the spec :D
And yes, will update.

gzh-terry replied:
Done.

jhedberg commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -124,12 +149,21 @@ struct bt_avrcp_subunit_info_rsp {
 
+#define BT_AVRCP_PASSTHROUGH_GET_STATE(payload)                                                    \
+	((bt_avrcp_button_state_t)(FIELD_GET(BIT(7), ((payload)->byte0))))
+#define BT_AVRCP_PASSTHROUGH_GET_OPID(payload)                                                     \
```
Here you're already using "op" instead of "operation", so just be consistent with this everywhere.

gzh-terry replied:
Done.

jhedberg commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -124,12 +149,21 @@ struct bt_avrcp_subunit_info_rsp {
+	uint8_t operation_data[0];
+};
+
+struct bt_avrcp_get_capabilities_rsp {
```
`struct bt_avrcp_get_cap_rsp`

gzh-terry replied:
Done.

jhedberg commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -16,6 +16,31 @@
+#define BT_AVRCP_COMPANY_ID_BLUETOOTH_SIG 0x001958
+
+/** @brief AVRCP Capability ID */
+typedef enum __packed {
+	BT_AVRCP_CAP_COMPANY_ID = 0x2,
+	BT_AVRCP_CAP_EVENTS_SUPPORTED = 0x3,
+} bt_avrcp_cap_t;
```
Here you're already using "cap", so just be consistent with this everywhere

gzh-terry replied:
Done.

jhedberg commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -243,22 +243,105 @@ static void avrcp_disconnected(struct bt_avctp *session)
+{
+	struct bt_avrcp_avc_pdu *pdu;
+	struct bt_avrcp_get_capabilities_rsp *rsp;
+	uint16_t length;
+	uint16_t expected_length;
```
```c
uint16_t len, expected_len;
```

gzh-terry replied:
Done.

jhedberg commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -306,37 +390,33 @@ static void avrcp_pass_through_handler(struct bt_avrcp *avrcp, struct net_buf *b
+	bt_avrcp_rsp_t result;
+
+	avrcp_hdr = net_buf_pull_mem(buf, sizeof(*avrcp_hdr));
 
 	if (cr == BT_AVCTP_CMD) {
 		/* ToDo */
 	} else { /* BT_AVCTP_RESPONSE */
 		if ((avrcp_cb != NULL) && (avrcp_cb->subunit_info_rsp != NULL)) {
-			avrcp_hdr = (struct bt_avrcp_header *)buf->data;
-			net_buf_pull(buf, sizeof(*avrcp_hdr));
 			if (buf->len < 2) {
 				LOG_ERR("Invalid passthrough length: %d", buf->len);
 				return;
 			}
 
-			rsp.response = BT_AVRCP_HDR_GET_CTYPE_OR_RSP(avrcp_hdr);
-			tmp = net_buf_pull_u8(buf);
-			rsp.operation_id = FIELD_GET(GENMASK(6, 0), tmp);
-			rsp.state = FIELD_GET(GENMASK(7, 7), tmp);
-			rsp.len = net_buf_pull_u8(buf);
-			rsp.payload = rsp.len ? buf->data : NULL;
+			result = BT_AVRCP_HDR_GET_CTYPE_OR_RSP(avrcp_hdr);
+			rsp = (struct bt_avrcp_passthrough_rsp *)buf->data;
 
-			avrcp_cb->passthrough_rsp(avrcp, &rsp);
+			avrcp_cb->passthrough_rsp(avrcp, result, rsp);
 		}
 	}
 }
 
 static const struct avrcp_handler handler[] = {
-	{ BT_AVRCP_OPC_VENDOR_DEPENDENT, avrcp_vendor_dependent_handler },
-	{ BT_AVRCP_OPC_UNIT_INFO, avrcp_unit_info_handler },
-	{ BT_AVRCP_OPC_SUBUNIT_INFO, avrcp_subunit_info_handler },
-	{ BT_AVRCP_OPC_PASS_THROUGH, avrcp_pass_through_handler },
+	{BT_AVRCP_OPC_VENDOR_DEPENDENT, avrcp_vendor_dependent_handler},
+	{BT_AVRCP_OPC_UNIT_INFO, avrcp_unit_info_handler},
+	{BT_AVRCP_OPC_SUBUNIT_INFO, avrcp_subunit_info_handler},
+	{BT_AVRCP_OPC_PASS_THROUGH, avrcp_pass_through_handler},
 };
```
Why is this change needed? clang-format might complain something, but the existing format was fine according to Zephyr coding style. One thing that could help with clang-format (I think) is to add a `,` ever after the handler pointer.

gzh-terry replied:
Reverted, there is an auto clang correction.

jhedberg commented on subsys/bluetooth/host/classic/avrcp_internal.h:
```diff
@@ -37,20 +38,95 @@ typedef enum __packed {
+
+struct bt_avrcp_avc_pdu {
+	uint8_t pdu_id;
+	uint8_t byte1;             /**< [7:2]: Reserved, [1:0]: Packet Type */
```
Just call this `pkt_type` or `type` instead of `byte1` for better readability.

gzh-terry replied:
Then a future developer may use the entire byte as packet type, however, only 2 bits shall be used.
Do we care more about readability or a risk of misuse?

jhedberg replied:
As long as the current code always accesses this member through a separate macro, any new developer should see this. They should also be aware of the AVRCP specification if they're touching this code. I don't have a super strong opinion on this, but especially since it's an internal header, it'll be easy to rename this later if a newer spec takes into use the reserved bits. (that said, I don't think the Bluetooth SIG is doing almost any updates for BR/EDR profiles anymore)

gzh-terry replied:
Done.

jhedberg commented on subsys/bluetooth/host/classic/avrcp_internal.h:
```diff
@@ -37,20 +38,95 @@ typedef enum __packed {
+struct bt_avrcp_avc_pdu {
+	uint8_t pdu_id;
+	uint8_t byte1;             /**< [7:2]: Reserved, [1:0]: Packet Type */
+	uint16_t parameter_length; /**< Use sys_cpu_to_be16() on write, sys_be16_to_cpu() on read */
+	uint8_t parameter[0];
```
```c
uint16_t param_len; /* big endian */
uint8_t param[0];
```

Actually, I'm not sure you should put any code comment there at all, since then we should do it everywhere for all protocol structs (and we don't do it anywhere currently).

gzh-terry replied:
The important thing is, the Core Spec does not always use big endian in these fields. Note that the most datafields are little endian in the Controller.
For someone doing both Host and Controller development (like me). This commend would be really helpful.

jhedberg replied:
Is the whole AVRCP/AVCTP protocol big endian? I agree, if this specific field is an exception, then a comment is justified. If you're making a general rule for big endian protocols we should be consistent across the board for those. E.g. SDP and OBEX are both big endian.

gzh-terry replied:
Typically those protocols above L2CAP are big endian. I would remove the comment to keep consistent.

gzh-terry replied:
As we do not have an assumption that our system is in little endian, at least one of `sys_cpu_to_be16` and `sys_cpu_to_le16` shall be used. Then a developer would have the motivation (and responsibility) to figure out which one is being used.
So yes, it is not necessary to have a comment here.

jhedberg reviewed (changes requested):
Some more cosmetic/style related things, but since there are parts which affect the public API I'd like to get this sorted out before merging.

gzh-terry commented on subsys/bluetooth/host/classic/avrcp_internal.h:
```diff
@@ -37,20 +38,95 @@ typedef enum __packed {
+
+	/** Error Response */
+	BT_AVRCP_PDU_ID_GENERAL_REJECT = 0xa0,
+} bt_avrcp_pdu_id_t;
```
Hi @jhedberg 

Do we need shortened macros as well?

Currently I'm using `BT_AVRCP_PDU_ID_##<the original name in the spec>`

jhedberg replied:
The code comments themselves give the reader an idea of which section in the spec these are defined, so I don't think there's risk of confusion if you do some shortening.

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/avrcp.c:
```diff
@@ -196,10 +218,8 @@ static int cmd_passthrough(const struct shell *sh, bt_avrcp_opid_t operation_id,
-				     len);
-		bt_avrcp_passthrough(default_avrcp, operation_id, BT_AVRCP_BUTTON_RELEASED, payload,
-				     len);
+		bt_avrcp_passthrough(default_avrcp, opid, BT_AVRCP_BUTTON_PRESSED, payload, len);
+		bt_avrcp_passthrough(default_avrcp, opid, BT_AVRCP_BUTTON_RELEASED, payload, len);
```
I have a question about the req-rsp model of AVRCP. After sending one command, whether we need to wait for rsp before sending the next one? If not, then does each command need its own timer to confirm the command waiting for the response timeout? If yes, is there any issue if you don't wait for the reply here?

gzh-terry replied:
> I have a question about the req-rsp model of AVRCP. After sending one command, whether we need to wait for rsp before sending the next one? If not, then does each command need its own timer to confirm the command waiting for the response timeout? If yes, is there any issue if you don't wait for the reply here?

Hi @lylezhu2012 ,

Yes, literally each command has its own timeout.

However, in this implementation, the timeout is only used for a warning log.
Alternatively, we can disconnect the AVRCP link as an error handling (in this case, just keep one timer is enough, since we don't really need a exact time duration).

gzh-terry replied:
Hi @lylezhu2012 

I've removed the timeout mechanism (and the TID things) in the latest patch.
I've also discovered that we have no cases that really care about the TID at the App side (for CT role). Therefore, no API changes have been made here.
Would you mind taking a look to see if this is acceptable to you?

Best regards,

gzh-terry commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -589,6 +689,26 @@ static int avrcp_send(struct bt_avrcp *avrcp, struct net_buf *buf)
+	}
+
+	net_buf_add_be24(buf, BT_AVRCP_COMPANY_ID_BLUETOOTH_SIG);
+	pdu.pdu_id = BT_AVRCP_PDU_ID_GET_CAPS;
+	BT_AVRCP_AVC_PDU_SET_PACKET_TYPE(&pdu, BT_AVRVP_PKT_TYPE_SINGLE);
+	pdu.param_len = sys_cpu_to_be16(sizeof(cap_id));
+	net_buf_add_mem(buf, &pdu, sizeof(pdu));
```
Done.

gzh-terry committed:
Bluetooth: AVRCP: expose macros to upper layer.

This patch expose BT_AVRCP_COMPANY_ID_SIZE to applications.

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

gzh-terry commented on subsys/bluetooth/host/classic/avrcp_internal.h:
```diff
@@ -18,7 +18,8 @@
 #define AVRCP_SUBUNIT_PAGE           (0) /* Fixed value according to AVRCP */
-#define AVRCP_SUBUNIT_EXTENSION_COED (7) /* Fixed value according to TA Document 2001012 */
+#define AVRCP_SUBUNIT_EXTENSION_CODE (7) /* Fixed value according to TA Document 2001012 */
+#define AVRCP_COMPANY_ID_SIZE        (3) /* Length for each company ID */
```
Done.

gzh-terry commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -306,28 +390,24 @@ static void avrcp_pass_through_handler(struct bt_avrcp *avrcp, struct net_buf *b
 		if ((avrcp_cb != NULL) && (avrcp_cb->subunit_info_rsp != NULL)) {
-			avrcp_hdr = (struct bt_avrcp_header *)buf->data;
-			net_buf_pull(buf, sizeof(*avrcp_hdr));
 			if (buf->len < 2) {
```
Done.

gzh-terry committed:
Bluetooth: AVRCP: Remove AVRCP timeout.

Remove the transaction label mechanism because:
1. Not necessary at stack, typically an application can handle the
timeout mechanism if required (very rarely).
2. It is a burden on memory size to record all pending commands.

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

gzh-terry committed:
Bluetooth: AVRCP: Move the handling of transaction labels to the App.

According to AVCTP v1.4: On the controller and the target side,
handling of transaction labels is dependent on the application.

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

lylezhu2012 commented:
> We have discussed internally that the handling of `tid` (referred to as "transaction label" in AVCTP) has now been moved to the app side. Therefore, the application itself can determine how to act on the timeout mechanisms.

I think it looks good for me.

But I have another doubt here, whether is it a good idea that the timeout of command-response defined by AVRCP specification needs to be implemented by application rather than AVRCP?  @jhedberg Do you have any idea? Is there a similar case for Zephyr Bluetooth currently?

jhedberg commented:
> timeout of command-response defined by AVRCP specification needs to be implemented by application rather than AVRCP? @jhedberg Do you have any comments? Is there a similar case for Zephyr Bluetooth currently?

No. All other protocols implement the timeout handling in the stack itself. My initial feeling is that it'd be better to follow the same design philosophy for AVRCP as well.

gzh-terry commented:
> > timeout of command-response defined by AVRCP specification needs to be implemented by application rather than AVRCP? @jhedberg Do you have any comments? Is there a similar case for Zephyr Bluetooth currently?
> 
> No. All other protocols implement the timeout handling in the stack itself. My initial feeling is that it'd be better to follow the same design philosophy for AVRCP as well.

Hi @jhedberg 

Yes, it would be better to keep it consistent with other protocols. We can always reintroduce the timeout mechanism when necessary. However, we haven't found a practical case that depends on the timeout at the AVCTP level. Therefore, I would prefer to keep it simple at this moment.
